runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -2984,48 +2986,46 @@ J9::ValuePropagation::isArrayCompTypePrimitiveValueType(TR::VPConstraint *arrayC
    if (!(arrayConstraint && arrayConstraint->getClass()
               && arrayConstraint->getClassType()->isArray() == TR_yes))
       {
+      if (trace())
+         traceMsg(comp(), "%s: return TR_maybe. arrayConstraint %p\n", __FUNCTION__, arrayConstraint);
       return TR_maybe;
       }
 
-   TR_OpaqueClassBlock *arrayComponentClass = fe()->getComponentClassFromArrayClass(arrayConstraint->getClass());
-
-   // Cases to consider:
-   //
-   //   - Is no information available about the component type of the array?
-   //     If not, assume it might be a primitive value type.
-   //   - Is the component type definitely a identity type?
-   //   - Is the component type definitely a primitive value type?
-   //   - Is the component type definitely a value type, but not primitive?
-   //   - Is the component type either an abstract class or an interface
-   //     (i.e., not a concrete class)?  If so, it might be a value type.
-   //   - Is the array an array of java/lang/Object?  See below.
-   //   - Otherwise, it must be a concrete class known not to be a value
-   //     type
-   //
-   if (!arrayComponentClass)
-      {
-      return TR_maybe;
-      }
-
-   // No need to check array class type because array classes should be marked as having identity.
-   if (TR::Compiler->cls.classHasIdentity(arrayComponentClass))
-      {
-      return TR_no;
-      }
+   TR_OpaqueClassBlock *arrayClass = arrayConstraint->getClass();
 
-   if (TR::Compiler->cls.isPrimitiveValueTypeClass(arrayComponentClass))
+   if (TR::Compiler->cls.isArrayNullRestricted(comp(), arrayClass))
       {
+      if (trace())
+         traceMsg(comp(), "%s: return TR_yes. arrayClass isArrayNullRestricted\n", __FUNCTION__);
```

Comment:
If the array's component type is a value type and its allocation is visible to the JIT compiler &mdash; for instance, it was allocated through the `anewarray` bytecode &mdash; it should be possible for Value Propagation to know that it is not a null-restricted array.  I'm wondering whether the `arrayConstraint` could/would be listed as a `fixedClass` in that case, and the result of `cls.isArrayNullRestricted` used to determine whether this method could return `TR_no`?

That is, given the following sequence of code

```
void method(ValueType[] parm) {
ValueType[] arr1 = ValueClass.newArrayInstance(NullRestrictedCheckedType.of(ValueType), 10);
VaueType[] arr2 = new ValueClass[10];
...
}
```

would it be possible for `vp->isArrayNullRestricted` to return `TR_maybe`, `TR_yes` and `TR_no` for `parm`, `arr1` and `arr2`, respectively?  I think as things stand, it will be return `TR_maybe`, `TR_yes` and `TR_maybe`.